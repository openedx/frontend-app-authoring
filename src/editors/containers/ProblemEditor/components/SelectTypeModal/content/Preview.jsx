import React from 'react';
import PropTypes from 'prop-types';
import { Hyperlink, Image, Container } from '@openedx/paragon';
import {
  FormattedMessage,
  useIntl,
} from '@edx/frontend-platform/i18n';
import messages from './messages';
import { ProblemTypes } from '../../../../../data/constants/problem';

const Preview = ({
  problemType,
}) => {
  const intl = useIntl();
  if (problemType === null) {
    return null;
  }
  const data = ProblemTypes[problemType];
  return (
    <Container style={{ width: '494px', height: '400px' }} className="bg-light-300 rounded p-4">
      <div className="small">
        {intl.formatMessage(messages.previewTitle, { previewTitle: data.title })}
      </div>
      <Image
        fluid
        className="my-3"
        src={data.preview}
        alt={intl.formatMessage(messages.previewAltText, { problemType })}
      />
      <div className="mb-3">
        {intl.formatMessage(messages.previewDescription, { previewDescription: data.previewDescription })}
      </div>
      <Hyperlink
        destination={data.helpLink}
        target="_blank"
      >
        <FormattedMessage {...messages.learnMoreButtonLabel} />
      </Hyperlink>
    </Container>
  );
};

Preview.defaultProps = {
  problemType: null,
};

Preview.propTypes = {
  problemType: PropTypes.string,
};

export default Preview;
