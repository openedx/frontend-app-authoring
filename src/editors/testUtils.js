/* istanbul ignore file */
import react from 'react';
import { StrictDict } from './utils';

/**
 * Mocked formatMessage provided by react-intl
 */
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) {
    return message;
  }
  Object.keys(values).forEach((key) => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};

/**
 * Mock a single component, or a nested component so that its children render nicely
 * in snapshots.
 * @param {string} name - parent component name
 * @param {obj} contents - object of child components with intended component
 *   render name.
 * @return {func} - mock component with nested children.
 *
 * usage:
 *   mockNestedComponent('Card', { Body: 'Card.Body', Form: { Control: { Feedback: 'Form.Control.Feedback' }}... });
 *   mockNestedComponent('IconButton', 'IconButton');
 */
export const mockNestedComponent = (name, contents) => {
  if (typeof contents !== 'object') {
    return contents;
  }
  const fn = () => name;
  Object.defineProperty(fn, 'name', { value: name });
  Object.keys(contents).forEach((nestedName) => {
    const value = contents[nestedName];
    fn[nestedName] = typeof value !== 'object'
      ? value
      : mockNestedComponent(`${name}.${nestedName}`, value);
  });
  return fn;
};

/**
 * Mock a module of components.  nested components will be rendered nicely in snapshots.
 * @param {obj} mapping - component module mock config.
 * @return {obj} - module of flat and nested components that will render nicely in snapshots.
 * usage:
 *   mockNestedComponents({
 *     Card: { Body: 'Card.Body' },
 *     IconButton: 'IconButton',
 *   })
 */
export const mockNestedComponents = (mapping) => Object.entries(mapping).reduce(
  (obj, [name, value]) => ({
    ...obj,
    [name]: mockNestedComponent(name, value),
  }),
  {},
);

/**
 * Mock utility for working with useState in a hooks module.
 * Expects/requires an object containing the state object in order to ensure
 * the mock behavior works appropriately.
 *
 * Expected format:
 *   hooks = { state: { <key>: (val) => React.createRef(val), ... } }
 *
 * Returns a utility for mocking useState and providing access to specific state values
 * and setState methods, as well as allowing per-test configuration of useState value returns.
 *
 * Example usage:
 *   // hooks.js
 *   import * as module from './hooks';
 *   const state = {
 *     isOpen: (val) => React.useState(val),
 *     hasDoors: (val) => React.useState(val),
 *     selected: (val) => React.useState(val),
 *   };
 *   ...
 *   export const exampleHook = () => {
 *     const [isOpen, setIsOpen] = module.state.isOpen(false);
 *     if (!isOpen) { return null; }
 *     return { isOpen, setIsOpen };
 *   }
 *   ...
 *
 *   // hooks.test.js
 *   import * as hooks from './hooks';
 *   const state = new MockUseState(hooks)
 *   ...
 *   describe('state hooks', () => {
 *     state.testGetter(state.keys.isOpen);
 *     state.testGetter(state.keys.hasDoors);
 *     state.testGetter(state.keys.selected);
 *   });
 *   describe('exampleHook', () => {
 *     beforeEach(() => { state.mock(); });
 *     it('returns null if isOpen is default value', () => {
 *       expect(hooks.exampleHook()).toEqual(null);
 *     });
 *     it('returns isOpen and setIsOpen if isOpen is not null', () => {
 *       state.mockVal(state.keys.isOpen, true);
 *       expect(hooks.exampleHook()).toEqual({
 *         isOpen: true,
 *         setIsOpen: state.setState[state.keys.isOpen],
 *       });
 *     });
 *     afterEach(() => { state.restore(); });
 *   });
 *
 * @param {obj} hooks - hooks module containing a 'state' object
 */
export class MockUseState {
  constructor(hooks) {
    this.hooks = hooks;
    this.oldState = null;
    this.setState = {};
    this.stateVals = {};

    this.mock = this.mock.bind(this);
    this.restore = this.restore.bind(this);
    this.mockVal = this.mockVal.bind(this);
    this.testGetter = this.testGetter.bind(this);
  }

  /**
   * @return {object} - StrictDict of state object keys
   */
  get keys() {
    return StrictDict(Object.keys(this.hooks.state).reduce(
      (obj, key) => ({ ...obj, [key]: key }),
      {},
    ));
  }

  /**
   * Replace the hook module's state object with a mocked version, initialized to default values.
   */
  mock() {
    this.oldState = this.hooks.state;
    Object.keys(this.keys).forEach(key => {
      this.hooks.state[key] = jest.fn(val => {
        this.stateVals[key] = val;
        return [val, this.setState[key]];
      });
    });
    this.setState = Object.keys(this.keys).reduce(
      (obj, key) => ({
        ...obj,
        [key]: jest.fn(val => {
          this.hooks.state[key] = val;
        }),
      }),
      {},
    );
  }

  /**
   * Restore the hook module's state object to the actual code.
   */
  restore() {
    this.hooks.state = this.oldState;
  }

  /**
   * Mock the state getter associated with a single key to return a specific value one time.
   * @param {string} key - state key (from this.keys)
   * @param {any} val - new value to be returned by the useState call.
   */
  mockVal(key, val) {
    this.hooks.state[key].mockReturnValueOnce([val, this.setState[key]]);
  }

  testGetter(key) {
    test(`${key} state getter should return useState passthrough`, () => {
      const testValue = 'some value';
      const useState = (val) => ({ useState: val });
      jest.spyOn(react, 'useState').mockImplementationOnce(useState);
      expect(this.hooks.state[key](testValue)).toEqual(useState(testValue));
    });
  }
}
