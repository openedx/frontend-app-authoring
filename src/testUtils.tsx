/* istanbul ignore file */
/* eslint-disable react/prop-types */
/* eslint-disable import/no-extraneous-dependencies */
/**
 * Helper functions for writing tests.
 */
import React from 'react';
import { AxiosError } from 'axios';
import { jest } from '@jest/globals';
import type { Store } from 'redux';
import { initializeMockApp } from '@edx/frontend-platform';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import { AppProvider } from '@edx/frontend-platform/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, type RenderResult } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';

import { ToastContext, type ToastContextData } from './generic/toast-context';
import initializeReduxStore from './store';

/** @deprecated Use React Query and/or regular React Context instead of redux */
let reduxStore: Store;
let queryClient;
let axiosMock: MockAdapter;

/** To use this: `const { mockShowToast } = initializeMocks()` and `expect(mockShowToast).toHaveBeenCalled()` */
let mockToastContext: ToastContextData = {
  showToast: jest.fn(),
  closeToast: jest.fn(),
  toastAction: undefined,
  toastMessage: null,
};

export interface RouteOptions {
  /** The URL path, like '/libraries/:libraryId' */
  path?: string;
  /** The URL parameters, like {libraryId: 'lib:org:123'} */
  params?: Record<string, string>;
  /** and/or instead of specifying path and params, specify MemoryRouterProps */
  routerProps?: MemoryRouterProps;
}

export interface WrapperOptions {
  extraWrapper?: React.FunctionComponent<{ children: React.ReactNode; }>;
}

/**
 * This component works together with the custom `render()` method we have in
 * this file to provide whatever react-router context you need for your
 * component.
 *
 * In the simplest case, you don't need to worry about the router at all, so
 * just render your component using `render(<TheComponent />)`.
 *
 * The next simplest way to use it is to specify `path` (the route matching rule
 * that is normally used to determine when to show the component or its parent
 * page) and `params` like this:
 *
 * ```
 * render(<LibraryLayout />, { path: '/library/:libraryId/*', params: { libraryId: 'lib:Axim:testlib' } });
 * ```
 *
 * In this case, components that use the `useParams` hook will get the right
 * library ID, and we don't even have to mock anything.
 *
 * In other cases, such as when you have routes inside routes, you'll need to
 * set the router's `initialEntries` (URL history) prop yourself, like this:
 *
 * ```
 * render(<LibraryLayout />, {
 *   path: '/library/:libraryId/*',
 *   // The root component is mounted on the above path, as it is in the "real"
 *   // MFE. But to access the 'settings' sub-route/component for this test, we
 *   // need tospecify the URL like this:
 *   routerProps: { initialEntries: [`/library/${libraryId}/settings`] },
 * });
 * ```
 */
const RouterAndRoute: React.FC<RouteOptions> = ({
  children,
  path = '/',
  params = {},
  routerProps = {},
}) => {
  if (Object.entries(params).length > 0 || path !== '/') {
    const newRouterProps = { ...routerProps };
    if (!routerProps.initialEntries) {
      // Substitute the params into the URL so '/library/:libraryId' becomes '/library/lib:org:123'
      let pathWithParams = path;
      for (const [key, value] of Object.entries(params)) {
        pathWithParams = pathWithParams.replaceAll(`:${key}`, value);
      }
      if (pathWithParams.endsWith('/*')) {
        // Some routes (that contain child routes) need to end with /* in the <Route> but not in the router
        pathWithParams = pathWithParams.substring(0, pathWithParams.length - 1);
      }
      newRouterProps.initialEntries = [pathWithParams];
    }
    return (
      <MemoryRouter {...newRouterProps}>
        <Routes>
          <Route path={path} element={children} />
        </Routes>
      </MemoryRouter>
    );
  }
  return (
    <MemoryRouter {...routerProps}>{children}</MemoryRouter>
  );
};

function makeWrapper({ extraWrapper, ...routeArgs }: WrapperOptions & RouteOptions) {
  const AllTheProviders = ({ children }) => (
    <AppProvider store={reduxStore} wrapWithRouter={false}>
      <IntlProvider locale="en" messages={{}}>
        <QueryClientProvider client={queryClient}>
          <ToastContext.Provider value={mockToastContext}>
            <RouterAndRoute {...routeArgs}>
              {extraWrapper ? React.createElement(extraWrapper, undefined, children) : children}
            </RouterAndRoute>
          </ToastContext.Provider>
        </QueryClientProvider>
      </IntlProvider>
    </AppProvider>
  );
  return AllTheProviders;
}

/**
 * Same as render() from `@testing-library/react` but this one provides all the
 * wrappers our React components need to render properly.
 */
function customRender(ui: React.ReactElement, options: WrapperOptions & RouteOptions = {}): RenderResult {
  return render(ui, { wrapper: makeWrapper(options) });
}

const defaultUser = {
  userId: 3,
  username: 'abc123',
  administrator: true,
  roles: [],
} as const;

/**
 * Initialize common mocks that many of our React components will require.
 *
 * This should be called within each test case, or in `beforeEach()`.
 *
 * Returns the new `axiosMock` in case you need to mock out axios requests.
 */
export function initializeMocks({ user = defaultUser, initialState = undefined }: {
  user?: { userId: number, username: string },
  initialState?: Record<string, any>, // TODO: proper typing for our redux state
} = {}) {
  initializeMockApp({
    authenticatedUser: user,
  });
  reduxStore = initializeReduxStore(initialState as any);
  queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });
  axiosMock = new MockAdapter(getAuthenticatedHttpClient());

  // Reset `mockToastContext` for this current test
  mockToastContext = {
    showToast: jest.fn(),
    closeToast: jest.fn(),
    toastMessage: null,
    toastAction: undefined,
  };

  // Clear the call counts etc. of all mocks. This doesn't remove the mock's effects; just clears their history.
  jest.clearAllMocks();

  return {
    reduxStore,
    axiosMock,
    mockShowToast: mockToastContext.showToast,
    mockToastAction: mockToastContext.toastAction,
  };
}

export * from '@testing-library/react';
export { customRender as render };

/** Simulate a real Axios error (such as we'd see in response to a 404) */
export function createAxiosError({ code, message, path }: { code: number, message: string, path: string }) {
  const request = { path };
  const config = {};
  const error = new AxiosError(
    `Mocked request failed with status code ${code}`,
    AxiosError.ERR_BAD_RESPONSE,
    config,
    request,
    {
      status: code,
      data: { detail: message },
      statusText: 'error',
      config,
      headers: {},
    },
  );
  return error;
}
