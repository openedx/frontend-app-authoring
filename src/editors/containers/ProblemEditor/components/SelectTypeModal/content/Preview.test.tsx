import React from 'react';
import { render, screen, initializeMocks } from '../../../../../../testUtils';
import Preview from './Preview';

// Mock ProblemTypes and messages
jest.mock('../../../../../data/constants/problem', () => ({
  ProblemTypes: {
    example: {
      title: 'Example Title',
      preview: 'example.png',
      previewDescription: 'Example description',
      helpLink: 'https://help.example.com',
    },
  },
}));

describe('Preview', () => {
  beforeEach(() => {
    initializeMocks();
  });

  it('renders nothing if problemType is null', () => {
    const { container } = render(<Preview problemType={null} />);
    const reduxProviderDiv = container.querySelector('div[data-testid="redux-provider"]');
    expect(reduxProviderDiv?.innerHTML).toBe('');
  });

  it('renders preview with correct data for a valid problemType', () => {
    render(<Preview problemType="example" />);
    expect(screen.getByText('Example Title problem')).toBeInTheDocument();
    expect(screen.getByText('Example description')).toBeInTheDocument();
    expect(screen.getByRole('img')).toHaveAttribute('src', 'example.png');
    expect(screen.getByRole('img')).toHaveAttribute('alt', 'A preview illustration of a null problem');
    expect(screen.getByRole('link', { name: 'Learn more in a new tab' })).toHaveAttribute('href', 'https://help.example.com');
  });

  it('renders the help link with target="_blank"', () => {
    render(<Preview problemType="example" />);
    const link = screen.getByRole('link', { name: 'Learn more in a new tab' });
    expect(link).toHaveAttribute('target', '_blank');
  });

  it('renders the correct title and description', () => {
    render(<Preview problemType="example" />);
    expect(screen.getByText('Example Title problem')).toBeInTheDocument();
    expect(screen.getByText('Example description')).toBeInTheDocument();
  });
});
