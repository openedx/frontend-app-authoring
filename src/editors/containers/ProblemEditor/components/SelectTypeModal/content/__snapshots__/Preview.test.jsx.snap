// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview snapshots snapshots: renders as expected with default props 1`] = `null`;

exports[`Preview snapshots snapshots: renders as expected with problemType is choiceresponse 1`] = `
<Container
  className="bg-light-300 rounded p-4"
  style={
    {
      "height": "400px",
      "width": "494px",
    }
  }
>
  <div
    className="small"
  >
    Multi-select problem
  </div>
  <Image
    alt="A preview illustration of a {problemType, select,
      multiplechoiceresponse {single select}
      stringreponse {text input}
      numericalresponse {numerical input}
      optionresponse {dropdown}
      choiceresponse {multiple select}
      other {null}
    } problem"
    className="my-3"
    fluid={true}
    src="test-file-stub"
  />
  <div
    className="mb-3"
  >
    Learners must select all correct answers from a list of possible options.
  </div>
  <Hyperlink
    destination="https://docs.openedx.org/en/latest/educators/how-tos/course_development/exercise_tools/add_multi_select.html"
    target="_blank"
  >
    <FormattedMessage
      defaultMessage="Learn more"
      description="Label for Learn more button"
      id="authoring.problemEditor.learnMoreButtonLabel.label"
    />
  </Hyperlink>
</Container>
`;

exports[`Preview snapshots snapshots: renders as expected with problemType is multiplechoiceresponse 1`] = `
<Container
  className="bg-light-300 rounded p-4"
  style={
    {
      "height": "400px",
      "width": "494px",
    }
  }
>
  <div
    className="small"
  >
    Single select problem
  </div>
  <Image
    alt="A preview illustration of a {problemType, select,
      multiplechoiceresponse {single select}
      stringreponse {text input}
      numericalresponse {numerical input}
      optionresponse {dropdown}
      choiceresponse {multiple select}
      other {null}
    } problem"
    className="my-3"
    fluid={true}
    src="test-file-stub"
  />
  <div
    className="mb-3"
  >
    Learners must select the correct answer from a list of possible options.
  </div>
  <Hyperlink
    destination="https://docs.openedx.org/en/latest/educators/concepts/exercise_tools/about_multi_select.html"
    target="_blank"
  >
    <FormattedMessage
      defaultMessage="Learn more"
      description="Label for Learn more button"
      id="authoring.problemEditor.learnMoreButtonLabel.label"
    />
  </Hyperlink>
</Container>
`;

exports[`Preview snapshots snapshots: renders as expected with problemType is numericalresponse 1`] = `
<Container
  className="bg-light-300 rounded p-4"
  style={
    {
      "height": "400px",
      "width": "494px",
    }
  }
>
  <div
    className="small"
  >
    Numerical input problem
  </div>
  <Image
    alt="A preview illustration of a {problemType, select,
      multiplechoiceresponse {single select}
      stringreponse {text input}
      numericalresponse {numerical input}
      optionresponse {dropdown}
      choiceresponse {multiple select}
      other {null}
    } problem"
    className="my-3"
    fluid={true}
    src="test-file-stub"
  />
  <div
    className="mb-3"
  >
    Specify one or more correct numeric answers, submitted in a response field.
  </div>
  <Hyperlink
    destination="https://docs.openedx.org/en/latest/educators/how-tos/course_development/exercise_tools/manage_numerical_input_problem.html"
    target="_blank"
  >
    <FormattedMessage
      defaultMessage="Learn more"
      description="Label for Learn more button"
      id="authoring.problemEditor.learnMoreButtonLabel.label"
    />
  </Hyperlink>
</Container>
`;

exports[`Preview snapshots snapshots: renders as expected with problemType is optionresponse 1`] = `
<Container
  className="bg-light-300 rounded p-4"
  style={
    {
      "height": "400px",
      "width": "494px",
    }
  }
>
  <div
    className="small"
  >
    Dropdown problem
  </div>
  <Image
    alt="A preview illustration of a {problemType, select,
      multiplechoiceresponse {single select}
      stringreponse {text input}
      numericalresponse {numerical input}
      optionresponse {dropdown}
      choiceresponse {multiple select}
      other {null}
    } problem"
    className="my-3"
    fluid={true}
    src="test-file-stub"
  />
  <div
    className="mb-3"
  >
    Learners must select the correct answer from a list of possible options
  </div>
  <Hyperlink
    destination="https://docs.openedx.org/en/latest/educators/how-tos/course_development/exercise_tools/add_dropdown.html"
    target="_blank"
  >
    <FormattedMessage
      defaultMessage="Learn more"
      description="Label for Learn more button"
      id="authoring.problemEditor.learnMoreButtonLabel.label"
    />
  </Hyperlink>
</Container>
`;

exports[`Preview snapshots snapshots: renders as expected with problemType is stringresponse 1`] = `
<Container
  className="bg-light-300 rounded p-4"
  style={
    {
      "height": "400px",
      "width": "494px",
    }
  }
>
  <div
    className="small"
  >
    Text input problem
  </div>
  <Image
    alt="A preview illustration of a {problemType, select,
      multiplechoiceresponse {single select}
      stringreponse {text input}
      numericalresponse {numerical input}
      optionresponse {dropdown}
      choiceresponse {multiple select}
      other {null}
    } problem"
    className="my-3"
    fluid={true}
    src="test-file-stub"
  />
  <div
    className="mb-3"
  >
    Specify one or more correct text answers, including numbers and special characters, submitted in a response field.
  </div>
  <Hyperlink
    destination="https://docs.openedx.org/en/latest/educators/how-tos/course_development/exercise_tools/add_text_input.html"
    target="_blank"
  >
    <FormattedMessage
      defaultMessage="Learn more"
      description="Label for Learn more button"
      id="authoring.problemEditor.learnMoreButtonLabel.label"
    />
  </Hyperlink>
</Container>
`;
